src/devices/video/upd7220.cpp

```diff
@@ -999,58 +985,155 @@ void upd7220_device::draw_rectangle(int x, int y)
 //  draw_char -
 //-------------------------------------------------
 
-void upd7220_device::draw_char(int x, int y)
+void upd7220_device::draw_char()
 {
-	int isize,psize;
-	uint16_t tile_data;
+	const int8_t dir_change[2][4] = {
+		{2, 2, -2, -2},
+		{1, 3, -3, -1}
+	};
+	const uint8_t type = (m_figs.m_figure_type & 0x10) >> 4;
 
-	LOG("uPD7220 char check: %d %d %02x %08x %d %d %02x\n",x,y,m_figs.m_dir,m_ead,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
+	uint16_t pattern;
 
-	/* QX10 may require upper bits for psize, VT240 requires the opposite */
-	isize = m_figs.m_d;
-	psize = m_figs.m_dc + 1;
+	LOG("uPD7220 char check: %08x %04x %02x %02x %d %d %02x\n",m_ead,m_mask,m_bitmap_mod,m_figs.m_dir,m_figs.m_d,m_figs.m_dc,m_figs.m_figure_type);
 
-	for(int pi = 0; pi < psize; pi++)
+	for (int i = 0, di = 0; i < (m_figs.m_dc + 1); ++i)
 	{
-		tile_data = (m_ra[15-(pi & 7)] << 8) | m_ra[15-(pi & 7)];
-		for(int pz = 0; pz <= m_gchr; pz++)
+		m_pattern = (m_ra[15 - (i & 7)] << 8) | m_ra[15 - (i & 7)];
```

Comment:
`m_pattern` is only used inside this loop, and doesn’t need its value to persist between iterations.  It shouldn’t be a member.  The `get_pattern` member function seems a little misguided, too.  Is there really a need to be checking the mode on every loop iteration?  You can lift the mode check to the top of this function and keep the result in a `const bool` local, make `m_pattern` a `const` local here, and then either expand the rest of `get_pattern` in-place using the ternary conditional operator, or make it a lambda.

Reply:
No this is here for future work. m_pattern represents the pattern register on the upd7220. This register is loaded from three sources PRAM(GCHRD, FIGD), WDAT, and DMAW. The get_pattern function then is the set of muxes that will take the output of the pattern register and give you the correct pattern to send to the logic unit based on the current mode and the RMW cycle. 

Right now this is the only place it is used, however once I get around to fixing up the FIGD commands and WDAT/DMAW i will be using both in those places as well.